### PR TITLE
Make app-ports not required for App Mesh

### PR DESCRIPTION
**Pull request by kiranmeduri**
*Issue #, if available:*

*Description of changes:*
Today appPorts and proxyIngressPort are required parameters in
aws-appmesh netconfig. However, for non-service tasks that do not listen
on any port, these fields are not necessary and ingress traffic
interception can be ignored. This change will make these two fields
optional but coupled, i.e. if one is provided then other should be
provided too.

Testing Done
* _Pending: Need to test in real-world scenario_
* Modified e2e_tests
```
CNI_PATH=./build/linux_amd64 go test -v -tags e2e_test -race -timeout 120s ./plugins/aws-appmesh/e2eTests/...
=== RUN   TestValid
=== RUN   TestValid/valid_ingress_egress
=== RUN   TestValid/valid_without_ingress
--- PASS: TestValid (1.42s)
    --- PASS: TestValid/valid_ingress_egress (0.65s)
    --- PASS: TestValid/valid_without_ingress (0.77s)
=== RUN   TestInvalid
=== RUN   TestInvalid/invalid_without_app_ports
=== RUN   TestInvalid/invalid_without_proxy_ingress_port
=== RUN   TestInvalid/invalid_without_proxy_egress_port
--- PASS: TestInvalid (0.45s)
    --- PASS: TestInvalid/invalid_without_app_ports (0.21s)
    --- PASS: TestInvalid/invalid_without_proxy_ingress_port (0.12s)
    --- PASS: TestInvalid/invalid_without_proxy_egress_port (0.13s)
PASS
ok      github.com/aws/amazon-vpc-cni-plugins/plugins/aws-appmesh/e2eTests      2.876s
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/aws-appmesh/config/netconfig.go [hunk 1]
```diff
@@ -40,8 +40,8 @@ type NetConfig struct {
 	EnableIPv6         bool
 }
 
-// netConfigJSON defines the network configuration JSON file format for the aws-appmesh cni plugin.
-type netConfigJSON struct {
+// NetConfigJSON defines the network configuration JSON file format for the aws-appmesh cni plugin.
+type NetConfigJSON struct {
```
Comment (ofiliz) [inline plugins/aws-appmesh/config/netconfig.go]: netConfigJSON is the internal representation of the JSON file, which is why it was private to the config package. Can we use the exported NetConfig structure in the e2etests instead? See my other comments below.

File: plugins/aws-appmesh/config/netconfig.go [hunk 2]
```diff
@@ -98,21 +98,26 @@ func New(args *cniSkel.CmdArgs) (*NetConfig, error) {
...
-	if len(config.AppPorts) == 0 {
-		return fmt.Errorf("missing required parameter appPorts")
-	}
+
 	if config.ProxyEgressPort == "" {
 		return fmt.Errorf("missing required parameter proxyEgressPort")
 	}
-	if config.ProxyIngressPort == "" {
+
+	// AppPorts and ProxyIngressPort go in pairs
```
Comment (yumex93) [inline plugins/aws-appmesh/config/netconfig.go]: nit: a couple of comments are missing period at the end. We are trying to make sure all comments are well formatted in this repo.

File: plugins/aws-appmesh/config/netconfig.go (hunk 1 above)
Comment (kiranmeduri) [reply plugins/aws-appmesh/config/netconfig.go]: IIUC it should be the other way. The input to CNI is json file and hence netconfigjson should be the interface (e.g. aws-appmesh.conf). e2e_test need to assert if the observed behavior against the json input. I could create a separate object to represent this json struct, but will end up creating netconfigjson again.

File: plugins/aws-appmesh/e2eTests/e2e_test.go [hunk 3]
```diff
@@ -171,90 +164,171 @@ func TestAddReturnError(t *testing.T) {
...
 		require.NoError(t, err, "Unable to initialize iptable")
 
-		exist, _ := iptable.Exists("nat", ingressChain, "-p", "tcp", "-m", "multiport", "--dports",
-			appPorts, "-j", "REDIRECT", "--to-port", proxyIngressPort)
-		require.False(t, exist, "Failed to delete rules to redirect app ports to proxy")
-
-		exist, _ = iptable.Exists("nat", "PREROUTING", "-p", "tcp", "-m", "addrtype", "!",
-			"--src-type", "LOCAL", "-j", ingressChain)
-		require.False(t, exist, "Failed to delete rules to jump from PREROUTING to ingress chain")
+		var conf config.NetConfigJSON
```
[the remark this answers is not in the document]
Comment (kiranmeduri) [reply plugins/aws-appmesh/e2eTests/e2e_test.go]: Yes they are used in all the tests. Inputs are loaded via testdata files and asserted in the tests. I have to cleanup the tests to get rid of the consts at the top and use the json struct, that will remove the confusion.

**Comment by yumex93** [comment]
Just wanna know, have you already asked ECS service team to make changes on validation accordingly?

**Comment by kiranmeduri** [comment]
> Just wanna know, have you already asked ECS service team to make changes on validation accordingly?

We cannot change the validation in the service layer given that it will be backward incompatible. But this change should work irrespective of ECS validation. Assuming this CNI can be used by itself in other places, I think it is okay to take this.

**Pull request closed without merging**